modules/apps/translation/translation-translator-deepl/src/main/java/com/liferay/translation/translator/deepl/internal/translator/DeepLClient.java

```diff
@@ -0,0 +1,144 @@
+
+/**
+ * @author Yasuyuki Takeo
+ */
+@Component(immediate = true, service = DeepLClient.class)
+public class DeepLClient {
+
+	public TranslateResponse execute(
+			String authKey, String text, String sourceLanguageId,
+			String targetLanguageId, String url)
+		throws IOException, TranslatorException {
+
+		return JSONUtil.toObject(
+			_fetch(authKey, text, sourceLanguageId, targetLanguageId, url),
+			TranslateResponse.class);
+	}
+
+	public List<SupportedLanguage> getSupportedLanguages(
+			String authKey, String target, String url)
+		throws IOException, TranslatorException {
+
+		return JSONUtil.toObject(
+			_fetchSupportedLanguage(authKey, target, url),
+			new TypeReference<List<SupportedLanguage>>() {
+			});
+	}
+
+	private String _fetch(
+			String authKey, String text, String sourceLanguageId,
+			String targetLanguageId, String url)
+		throws IOException, TranslatorException {
+
+		Http.Options options = new Http.Options();
+
+		options.setLocation(
+			URLBuilder.create(
+				url
+			).addParameter(
+				DeepLConstants.AUTH_KEY, authKey
+			).build());
+
+		options.addHeader(
+			HttpHeaders.CONTENT_TYPE,
+			ContentTypes.APPLICATION_X_WWW_FORM_URLENCODED);
+		options.addPart(DeepLConstants.AUTH_KEY, authKey);
+		options.addPart(DeepLConstants.SOURCE_LANG, sourceLanguageId);
+		options.addPart(DeepLConstants.TARGET_LANG, targetLanguageId);
+		options.addPart(DeepLConstants.TEXT, text);
+		options.setMethod(Http.Method.POST);
+
+		Http.Response response = options.getResponse();
+
+		Response.Status status = Response.Status.fromStatusCode(
+			response.getResponseCode());
+
+		if (status == Response.Status.OK) {
+			return _http.URLtoString(options);
+		}
+		else if (status == Response.Status.TOO_MANY_REQUESTS) {
+			throw new TranslatorException(
+				"Ths status is TOO_MANY_REQUESTS. Please retry after a while.");
+		}
+
+		return _http.URLtoString(options);
+	}
+
+	private String _fetchSupportedLanguage(
+			String authKey, String target, String url)
+		throws IOException, TranslatorException {
+
+		Http.Options options = new Http.Options();
+
+		options.setLocation(
+			URLBuilder.create(
+				url
+			).addParameter(
+				DeepLConstants.AUTH_KEY, authKey
+			).build());
+
+		options.addHeader(
+			HttpHeaders.CONTENT_TYPE,
+			ContentTypes.APPLICATION_X_WWW_FORM_URLENCODED);
+		options.addPart(DeepLConstants.AUTH_KEY, authKey);
+		options.addPart(DeepLConstants.TARGET, target);
+		options.setMethod(Http.Method.POST);
+
+		String supportedLanguage = _http.URLtoString(options);
+
+		Http.Response response = options.getResponse();
+
+		Response.Status status = Response.Status.fromStatusCode(
+			response.getResponseCode());
+
+		if (status == Response.Status.OK) {
+			return supportedLanguage;
+		}
+		else if (status == Response.Status.TOO_MANY_REQUESTS) {
+			throw new TranslatorException(
+				"Ths status is TOO_MANY_REQUESTS. Please retry after a while.");
+		}
+
+		return supportedLanguage;
```

Comment:
Shouldn't we fail here? If status is not OK you shouldn't return whatever the service returned and assume it is valid.